### PR TITLE
Feature/aapd 1032 create lpa questionnaire submission model

### PR DESCRIPTION
**Pull request by topickering**
## Ticket Number

https://pins-ds.atlassian.net/browse/AAPD-1032

## Description of change

Create LPASubmissionQuestionnaire sql model and associated models

## Checklist

- [ ] Requires infrastructure changes
- [ ] I have updated the documentation accordingly
- [ ] My commit history in this PR is linear
- [ ] New features have tests
- [ ] Breaking change (team conversation required)

## Important

Please do not merge from `main` (please only [rebase](https://github.com/foundry4/appeal-planning-decision/wiki/An-intro-to-Git-Rebase)). This keeps the history linear and easier to debug.

### CODE REVIEW

File: packages/appeals-service-api/src/db/schema.prisma [hunk 1]
```diff
@@ -1,5 +1,6 @@
 generator client {
-  provider = "prisma-client-js"
+  provider      = "prisma-client-js"
+  binaryTargets = ["native", "darwin"]
```
Comment (Si3rr4wow) [inline packages/appeals-service-api/src/db/schema.prisma:3]: I also had to add this a while ago but my prisma has been well behaved without it for a while. Do we definitively need this?

File: packages/appeals-service-api/src/db/schema.prisma [hunk 2]
```diff
@@ -266,3 +270,130 @@ model ServiceUser {
...
+  // Uploaded documents and details linked to this appeal
+  SubmissionDocumentUpload   SubmissionDocumentUpload[]
+  SubmissionNeighbourAddress SubmissionNeighbourAddress[]
+}
+
+model SubmissionDocumentUpload {
+  questionnaireId            String                     @db.UniqueIdentifier
+  LPAQuestionnaireSubmission LPAQuestionnaireSubmission @relation(fields: [questionnaireId], references: [id])
+
+  id               String @id @default(dbgenerated("newid()")) @db.UniqueIdentifier
```
Comment (Si3rr4wow) [inline packages/appeals-service-api/src/db/schema.prisma:382]: picky but I'd expect id first in a schema, might just be me though

File: packages/appeals-service-api/src/db/schema.prisma [hunk 3]
```diff
@@ -266,3 +270,130 @@ model ServiceUser {
...
+  name             String @default("")
+  fileName         String @default("")
+  originalFileName String @default("")
+  location         String @default("")
+  type             String @default("")
```
Comment (Si3rr4wow) [inline packages/appeals-service-api/src/db/schema.prisma:387]: Why not just leave these null?

File: packages/appeals-service-api/src/errors/apiError.js [hunk 4]
```diff
@@ -170,6 +170,42 @@ class ApiError {
...
+		return new ApiError(400, {
+			errors: ['Unable to create questionnaire']
+		});
+	}
+
+	static unableToGetDocumentUploads() {
+		return new ApiError(400, {
+			errors: ['Unable to retrieve document uploads']
+		});
+	}
```
Comment (Si3rr4wow) [inline packages/appeals-service-api/src/errors/apiError.js:189]: This feels 404esque, or is it not able to retrieve for a reason other that it couldn't be found?

File: packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/controller.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
+const {
+	getLPAQuestionnaireByAppealId,
+	createLPAQuestionnaire,
+	patchLPAQuestionnaireByAppealId
+} = require('./service');
+const logger = require('#lib/logger');
+const ApiError = require('#errors/apiError');
+
```
Comment (Si3rr4wow) [inline packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/controller.js:8]: I wonder if there's a way to dry this file up a bit but I might be overfactoring

File: packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/documentUpload/repo.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
+const { createPrismaClient } = require('#db-client');
+// const { PrismaClientKnownRequestError } = require('@prisma/client/runtime/library');
```
Comment (Si3rr4wow) [inline packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/documentUpload/repo.js:2]: 🗑️?

File: packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/documentUpload/repo.js [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+				questionnaireId
+			}
+		});
+	}
+
+	/**
+	 * Create submission document for given questionnaire
+	 *
+	 * @param {object} uploadData
+	 * @returns {Promise<SubmissionDocumentUpload|null>}
```
Comment (Si3rr4wow) [inline packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/documentUpload/repo.js:34]: Why `|null`?

File: packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/neighbourAddress/controller.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * @type {import('express').RequestHandler}
+ */
+async function createSubmissionNeighbourAddress(req, res) {
+	try {
+		// const { 'lpa-code': lpaCode } = req.query;
+		// if (!lpaCode || typeof lpaCode !== 'string') {
+		// 	throw ApiError.withMessage(400, 'lpa-code is required');
+		// }
```
Comment (Si3rr4wow) [inline packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/neighbourAddress/controller.js:38]: 🚮 ?

File: packages/forms-web-app/src/dynamic-forms/dynamic-components/address-add-more/question.js [hunk 9]
```diff
@@ -25,16 +28,24 @@ class AddressAddMoreQuestion extends AddMoreQuestion {
...
 		});
-		return { addMoreId: uuid.v4(), value: address };
+
+		await apiClient.postSubmissionNeighbourAddress(
+			journeyResponse.referenceId,
+			journeyResponse.answers.id,
+			address
+		);
+
+		return true;
```
Comment (Si3rr4wow) [inline packages/forms-web-app/src/dynamic-forms/dynamic-components/address-add-more/question.js:48]: Why?

File: packages/forms-web-app/src/dynamic-forms/dynamic-components/boolean/question.js [hunk 10]
```diff
@@ -37,6 +37,36 @@ class BooleanQuestion extends RadioQuestion {
...
 	}
+
+	/**
+	 * returns the data to send to the DB
+	 * side effect: modifies journeyResponse with the new answers
+	 * @param {ExpressRequest} req
+	 * @param {JourneyResponse} journeyResponse - current journey response, modified with the new answers
+	 * @returns {Promise.<Object>}
+	 */
+	async getDataToSave(req, journeyResponse) {
```
Comment (Si3rr4wow) [inline packages/forms-web-app/src/dynamic-forms/dynamic-components/boolean/question.js:48]: Wasn't expecting to see this file in this diff. How come this function didn't exist before?

File: packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js [hunk 11]
```diff
@@ -203,28 +204,45 @@ class MultiFileUploadQuestion extends Question {
...
+			const data = {
+				...file,
+				type: this.documentType.name
+			};
+			apiClient.postSubmissionDocumentUpload(
+				journeyResponse.referenceId,
+				journeyResponse.answers.id,
+				data
+			);
+		});
```
Comment (Si3rr4wow) [inline packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js:239]: is `apiClient. postSubmissionDocumentUpload` async?

File: packages/forms-web-app/src/dynamic-forms/middleware/get-journey-response-for-lpa.js [hunk 12]
```diff
@@ -20,10 +21,12 @@ module.exports = () => async (req, res, next) => {
...
 
 	try {
-		const dbResponse = await getQuestionResponse(appealType, encodedReferenceId);
-		result = new JourneyResponse(appealType, referenceId, dbResponse?.answers, dbResponse.LPACode);
+		const dbResponse = await apiClient.getLPAQuestionnaire(referenceId);
+		const convertedResponse = mapDBResponseToJourneyResponseFormat(dbResponse);
+		result = new JourneyResponse(appealType, referenceId, convertedResponse, dbResponse.lpaCode);
 	} catch (err) {
 		logger.error(err);
+		await apiClient.postLPAQuestionnaire(referenceId, user.lpaCode);
```
Comment (Si3rr4wow) [inline packages/forms-web-app/src/dynamic-forms/middleware/get-journey-response-for-lpa.js:29]: I don't understand this part, can you explain please?

File: packages/forms-web-app/src/dynamic-forms/middleware/utils.js [hunk 13]
```diff
@@ -30,4 +31,24 @@ function getAddMoreIfPresent(req, questionObj) {
...
+			convertedResponse[key] = 'yes';
+		} else if (dbResponse[key] === false) {
+			convertedResponse[key] = 'no';
+		} else {
+			convertedResponse[key] = dbResponse[key];
+		}
+	}
+
+	return convertedResponse;
+}
```
Comment (Si3rr4wow) [inline packages/forms-web-app/src/dynamic-forms/middleware/utils.js:52]:
```suggestion
function mapDBResponseToJourneyResponseFormat(dbResponse) {
	return Object.entries(dbResponse).reduce(
		(acc, [key, value]) => ({
			...acc,
			[key]: (() => {
				switch (value) {
					case true:
						return 'yes';
					case false:
						return 'no';
					default:
						return value;
				}
			})()
		}),
		{}
	);
}

```

feel free to not do this, I just can't help myself

File: packages/forms-web-app/src/dynamic-forms/question.js [hunk 14]
```diff
@@ -1,5 +1,6 @@
-const { patchQuestionResponse } = require('../lib/appeals-api-wrapper');
+// const { patchQuestionResponse } = require('../lib/appeals-api-wrapper');
```
Comment (Si3rr4wow) [inline packages/forms-web-app/src/dynamic-forms/question.js:1]: 🚮?

File: packages/forms-web-app/src/dynamic-forms/question.js [hunk 15]
```diff
@@ -244,16 +245,10 @@ class Question {
 
 	/**
 	 * @param {JourneyResponse} journeyResponse
-	 * @param {Promise<void>} responseToSave
+	 * @param {object} responseToSave
```
Comment (Si3rr4wow) [inline packages/forms-web-app/src/dynamic-forms/question.js:248]: Looks like this still returns a promise, maybe `Promise<*>` if `patchLPAQuestionnaire` has an unknown response

File: packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/documentUpload/controller.js
Comment (beejjacobs) [inline packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/documentUpload/controller.js:1]: Would prefer we keep with `hypenated-names` for routes, e.g. `document-upload`

File: packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/neighbourAddress/controller.js
Comment (beejjacobs) [inline packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/neighbourAddress/controller.js:1]: As per docs, would prefer `neighbour-address`

File: packages/common/src/client/appeals-api-client.js [hunk 16]
```diff
@@ -164,10 +165,97 @@ class AppealsApiClient {
...
+
+	/**
+	 * @param {string} caseReference
+	 * @param {string} lpaCode
+	 * @returns {Promise<(LPAQuestionnaireSubmission)>}
+	 */
+	async postLPAQuestionnaire(caseReference, lpaCode) {
+		const urlParams = new URLSearchParams();
+		urlParams.append('lpa-code', lpaCode);
+		const endpoint = `${v2}/appeal-cases/${caseReference}/lpa-questionnaire-submission?${urlParams.toString()}`;
```
Comment (beejjacobs) [inline packages/common/src/client/appeals-api-client.js:186]: If lpa-code is part of this API, it's not in the spec. Is this needed? There is only ever one submission per case.

File: packages/common/src/client/appeals-api-client.js [hunk 17]
```diff
@@ -164,10 +165,97 @@ class AppealsApiClient {
...
+	}
+
+	/**
+	 * @param {string} caseReference
+	 * @param {string} questionnaireId
+	 */
+	async getSubmissionDocumentUploads(caseReference, questionnaireId) {
+		const urlParams = new URLSearchParams();
+		urlParams.append('questionnaire-id', questionnaireId);
+		const endpoint = `${v2}/appeal-cases/${caseReference}/lpa-questionnaire-submission/documentUpload?${urlParams.toString()}`;
```
Comment (beejjacobs) [inline packages/common/src/client/appeals-api-client.js:208]: As with `lpa-code`, not sure this is needed. If we've got case ref, there is a 1:1 with questionnaire.

File: packages/forms-web-app/src/dynamic-forms/dynamic-components/address-add-more/question.js [hunk 18]
```diff
@@ -25,16 +28,24 @@ class AddressAddMoreQuestion extends AddMoreQuestion {
...
 			townCity: req.body[this.fieldName + '_townCity'],
 			postcode: req.body[this.fieldName + '_postcode']
 		});
-		return { addMoreId: uuid.v4(), value: address };
+
+		await apiClient.postSubmissionNeighbourAddress(
+			journeyResponse.referenceId,
+			journeyResponse.answers.id,
+			address
+		);
```
Comment (beejjacobs) [inline packages/forms-web-app/src/dynamic-forms/dynamic-components/address-add-more/question.js:46]: Is this the right place? The existing function just returns some data, it doesn't save it?

File: packages/forms-web-app/src/dynamic-forms/dynamic-components/boolean/question.js (hunk 10 above)
Comment (beejjacobs) [reply packages/forms-web-app/src/dynamic-forms/dynamic-components/boolean/question.js:48]: I'd guess mapping 'yes'|'no' to true/false.

**Review by beejjacobs — COMMENTED** [review]
Quite a few comments, sorry - but it is a lot of code!

Good work on this, not enviable having to tackle it.

File: packages/appeals-service-api/src/db/schema.prisma (hunk 1 above)
Comment (topickering) [reply packages/appeals-service-api/src/db/schema.prisma:3]: I'll take it out

File: packages/appeals-service-api/src/db/schema.prisma (hunk 2 above)
Comment (topickering) [reply packages/appeals-service-api/src/db/schema.prisma:382]: Moved

File: packages/appeals-service-api/src/errors/apiError.js (hunk 4 above)
Comment (topickering) [reply packages/appeals-service-api/src/errors/apiError.js:189]: yep, changed to 404

File: packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/neighbourAddress/controller.js
Comment (topickering) [reply packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/neighbourAddress/controller.js:1]: Changed to hyphenated

File: packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/documentUpload/repo.js (hunk 6 above)
Comment (topickering) [reply packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/documentUpload/repo.js:2]: binning is winning

File: packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/neighbourAddress/controller.js (hunk 8 above)
Comment (topickering) [reply packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/neighbourAddress/controller.js:38]: Binning = winning

File: packages/forms-web-app/src/dynamic-forms/question.js (hunk 14 above)
Comment (topickering) [reply packages/forms-web-app/src/dynamic-forms/question.js:1]: binned

File: packages/forms-web-app/src/dynamic-forms/middleware/utils.js (hunk 13 above)
Comment (topickering) [reply packages/forms-web-app/src/dynamic-forms/middleware/utils.js:52]: No, you're right, made the change

File: packages/forms-web-app/src/dynamic-forms/middleware/get-journey-response-for-lpa.js (hunk 12 above)
Comment (topickering) [reply packages/forms-web-app/src/dynamic-forms/middleware/get-journey-response-for-lpa.js:29]: getLPAQuestionnaire will error if there is no questionnaire associated with that appeal case reference, so on catching the error this hits the post endpoint to create a new db entry (ie questionnaire) linked to that case ref.  In practice, would only expect this to be used when first looking to access a questionnaire from the dashboard (ie going to the task list).

File: packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js (hunk 11 above)
Comment (topickering) [reply packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js:239]: yep, have added an await

File: packages/forms-web-app/src/dynamic-forms/dynamic-components/boolean/question.js (hunk 10 above)
Comment (topickering) [reply packages/forms-web-app/src/dynamic-forms/dynamic-components/boolean/question.js:48]: Exactly.  The superclass Question has a getDataToSave function which (broadly) extracts the value input on the relevant question page, sets it to be saved and updates the journeyResponse object.  BooleanQuestions return strings from the front end (I think it's a nunjucks thing) so we need to override the default getDataToSave and convert to a boolean.  At one stage I had the logic in the main question.js file, but thought it made more sense to extract it out to different types where the corresponding field in the schema doesn't match the value returned from the front end

File: packages/forms-web-app/src/dynamic-forms/options-question.js [hunk 19]
```diff
@@ -125,6 +125,38 @@ class OptionsQuestion extends Question {
...
 	}
+
+	/**
+	 * returns the data to send to the DB
+	 * side effect: modifies journeyResponse with the new answers
+	 * @param {ExpressRequest} req
+	 * @param {JourneyResponse} journeyResponse - current journey response, modified with the new answers
+	 * @returns {Promise.<Object>}
+	 */
+	async getDataToSave(req, journeyResponse) {
```
[the remark this answers is not in the document]
Comment (topickering) [reply packages/forms-web-app/src/dynamic-forms/options-question.js:136]: Same as with the BooleanQuestion - this is overriding the default getDataToSave for the Question class. For an OptionsQuestion, the front end returns a string array.  Prisma doesn't support String[] when using sqlLite, so this converts it to a comma separated string.  I'm open to any suggestions for another type to save as in the sql db.

File: packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/documentUpload/repo.js (hunk 7 above)
Comment (topickering) [reply packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/documentUpload/repo.js:34]: Good point, only returns null on a `find` so have removed from the create

File: packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/repo.js [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+		this.dbClient = createPrismaClient();
+	}
+
+	/**
+	 * Get questionnaire for given appeal
+	 *
+	 * @param {string} caseReference
+	 * @returns {Promise<LPAQuestionnaireSubmission|null>}
+	 */
+	async getLPAQuestionnaireByAppealId(caseReference) {
```
[the remark this answers is not in the document]
Comment (topickering) [reply packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/repo.js:21]: Yep, amended

File: packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/service.js [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+ * @param {string} appealCaseId
+ * @param {LPAQuestionnaireSubmission} data
+ * @return {Promise<LPAQuestionnaireSubmission|null>}
+ */
+async function patchLPAQuestionnaireByAppealId(appealCaseId, data) {
+	const questionnaire = repo.patchLPAQuestionnaireByAppealId(appealCaseId, data);
+
+	if (!questionnaire) {
+		return null;
+	}
```
[the remark this answers is not in the document]
Comment (topickering) [reply packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/service.js:54]: Good point, only returns null for a findUnique/ findMany so have removed the null for create / update calls

File: packages/forms-web-app/src/dynamic-forms/question.js (hunk 15 above)
Comment (topickering) [reply packages/forms-web-app/src/dynamic-forms/question.js:248]: Done

File: packages/appeals-service-api/src/db/schema.prisma [hunk 22]
```diff
@@ -266,3 +270,130 @@ model ServiceUser {
...
 }
+
+/// LPAQuestionnaireSubmission represents the questionnaire responses from an LPAUser
+model LPAQuestionnaireSubmission {
+  id String @id @default(dbgenerated("newid()")) @db.UniqueIdentifier
+
+  appealCaseReference String     @unique
+  AppealCase          AppealCase @relation(fields: [appealCaseReference], references: [caseReference])
+
+  lpaCode String
```
[the remark this answers is not in the document]
Comment (topickering) [reply packages/appeals-service-api/src/db/schema.prisma:281]: Good point, will remove

File: packages/common/src/client/appeals-api-client.js (hunk 16 above)
Comment (topickering) [reply packages/common/src/client/appeals-api-client.js:186]: Following the refactoring and understanding how to use prisma client a bit more I've taken this out

File: packages/forms-web-app/src/dynamic-forms/dynamic-components/address-add-more/question.js (hunk 18 above)
Comment (topickering) [reply packages/forms-web-app/src/dynamic-forms/dynamic-components/address-add-more/question.js]: moved

File: packages/common/src/client/appeals-api-client.js (hunk 17 above)
Comment (topickering) [reply packages/common/src/client/appeals-api-client.js]: Agreed, taken this out

File: packages/appeals-service-api/src/db/schema.prisma [hunk 23]
```diff
@@ -266,3 +270,130 @@ model ServiceUser {
...
+
+  // otherDesignationDetails         String?
+  // completedEnvironmentalStatement Boolean?
+  // appealSiteAccess                Boolean?
+  // appealSiteAccessDetails							String?
+  // lpaStatement                    String?
+  // lpaStatementDocuments           Boolean?
+  // lpaFinalComment                 Boolean?
+  // lpaFinalCommentDetails          String?
+  // lpaWitnesses                    Boolean?
```
[the remark this answers is not in the document]
Comment (topickering) [reply packages/appeals-service-api/src/db/schema.prisma]: Removed the commented fields

File: packages/appeals-service-api/src/db/migrations/20231120091203_users/migration.sql
Comment (beejjacobs) [inline packages/appeals-service-api/src/db/migrations/20231120091203_users/migration.sql:1]: Oops, I think you deleted too many migrations! The existing ones need to stay, as some of them are already deployed. We just wanted one new migration for the LPAQ changes.

File: packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js [hunk 24]
```diff
@@ -84,15 +85,62 @@ class MultiFileUploadQuestion extends Question {
...
+
+		// save
+		const { uploadedFiles } = await this.getDataToSave(req, journeyResponse);
+		await uploadedFiles.forEach(async (file) => {
+			const data = {
+				...file,
+				type: this.documentType.name
+			};
+			await apiClient.postSubmissionDocumentUpload(journeyResponse.referenceId, data);
+		});
```
Comment (beejjacobs) [inline packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js:125]: Let's go with something like:

`await Promise.all(uploadedFiles.map((file) => {
  const data = {
    ...file,
    type: this.documentType.name
  };
  return apiClient.postSubmissionDocumentUpload(journeyResponse.referenceId, data);
});
```

we don't need to wait for each in turn.

Possible future improvement: the API could accept an array.

File: packages/forms-web-app/src/dynamic-forms/middleware/get-journey-response-for-lpa.js [hunk 25]
```diff
@@ -20,10 +21,17 @@ module.exports = () => async (req, res, next) => {
...
+		const convertedResponse = mapDBResponseToJourneyResponseFormat(dbResponse);
+		result = new JourneyResponse(
+			appealType,
+			referenceId,
+			convertedResponse,
+			dbResponse.AppealCase?.LPACode
+		);
 	} catch (err) {
 		logger.error(err);
+		await apiClient.postLPAQuestionnaire(referenceId);
```
Comment (beejjacobs) [inline packages/forms-web-app/src/dynamic-forms/middleware/get-journey-response-for-lpa.js:34]: I think if we need to create the questionnaire when not found this should be more explicit. If the error is something else, we don't always want to send this post.

**Review by beejjacobs — CHANGES_REQUESTED** [review]
Few more points, but looking good!

File: packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js (hunk 24 above)
Comment (beejjacobs) [reply packages/forms-web-app/src/dynamic-forms/dynamic-components/multi-file-upload/question.js]: See https://github.com/Planning-Inspectorate/appeal-planning-decision/pull/3404/commits/88d69b26cace909c16a75a6b7cf09898cb96dd38

File: packages/forms-web-app/src/dynamic-forms/middleware/get-journey-response-for-lpa.js (hunk 25 above)
Comment (beejjacobs) [reply packages/forms-web-app/src/dynamic-forms/middleware/get-journey-response-for-lpa.js]: See https://github.com/Planning-Inspectorate/appeal-planning-decision/pull/3404/commits/4e8b3884aaedee71aeabe624e45a1079a8a06b17

File: packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/service.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (Si3rr4wow) [reply packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/service.js:53]: Looks like the type was updated, do we still need this if and the one in the function above?

File: packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/repo.js [hunk 26]
```diff
@@ -0,0 +1,102 @@
...
+	 * @param {*} caseReference
+	 * @param {*} data
+	 * @returns {Promise<LPAQuestionnaireSubmission>}
+	 */
+	async patchLPAQuestionnaireByAppealId(caseReference, data) {
+		return await this.dbClient.lPAQuestionnaireSubmission.update({
+			where: {
+				appealCaseReference: caseReference
+			},
+			data,
```
Comment (Si3rr4wow) [inline packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/repo.js:90]: Looks like this data just gets passed in from the service, is there anywhere that checks this data looks like what we expect or will Prisma complain if that happens?

**Review by Si3rr4wow — APPROVED** [review]
LGTM ⭐️ one question but it's probably not important

File: packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/repo.js (hunk 26 above)
Comment (beejjacobs) [reply packages/appeals-service-api/src/routes/v2/appeal-cases/_caseReference/lpa-questionnaire-submission/repo.js:90]: Ah, good spot. yes Prisma does complain - but he API then returned 500 which isn't right. I've handled that 400 case now, and added PATCH to the spec.

[See https://github.com/Planning-Inspectorate/appeal-planning-decision/pull/3404/commits/4e8fa0b65ec136edab9146982e2baba56120214d](https://github.com/Planning-Inspectorate/appeal-planning-decision/pull/3404/commits/8dbf0a55eb3f90df2fe9bbcabc4e77b3a2f828f3)

**Review by Si3rr4wow — APPROVED** [review]

**Review by beejjacobs — APPROVED** [review]
Changes complete.

**Pull request merged**